71 password recovery

MajoritySky2496 opened the pull request:
Made a password recovery screen

MajoritySky2496 committed:
71-Password-Recovery
Made:
1. api
2. response
3. dto
4. DataMapper
5. Exception mapper

MajoritySky2496 committed:
71-Password-Recovery
Made:
1. di
2. Repository
3. Interactor
4. DomainMapper
5. Data
6. State
7.ExceptionToErrorMapper

MajoritySky2496 committed:
71-Password-Recovery
Made:
1. PasswordRecoveryViewModel
2. State

MajoritySky2496 committed:
Merge branch 'dev' of https://github.com/CashAdv/CashADV into 71-Password-Recovery

MajoritySky2496 committed:
Merge branch 'feature/51-code-confirmation-view' of https://github.com/CashAdv/CashADV into 71-Password-Recovery

MajoritySky2496 committed:
71-Password-Recovery
Made:

MajoritySky2496 committed:
71-Password-Recovery

MajoritySky2496 committed:
71-Password-Recovery

MajoritySky2496 committed:
71-Password-Recovery

MajoritySky2496 committed:
71-Password-Recovery

MajoritySky2496 committed:
Merge branch 'dev' of https://github.com/CashAdv/CashADV into 71-Password-Recovery

MajoritySky2496 committed:
71-Password-Recovery

MajoritySky2496 committed:
71-Password-Recovery

MajoritySky2496 committed:
71-Password-Recovery

MajoritySky2496 commented:
Есть одна проблема которую не могу решить, при вводе неверного когда приходить ошибка "Something went wrong Field 'message' is required for type with serial name 'app.cashadvisor.authorization.data.models.response.ConfirmResetPasswordResponse', but it was missing at path: $" Хотя данное поле существует. При верном вводе кода все работает и никаких ошибок нет. Буду благодарен за помощь)

MajoritySky2496 committed:
71-Password-Recovery

artwist-polyakov commented:
> Есть одна проблема которую не могу решить, при вводе неверного когда приходить ошибка "Something went wrong Field 'message' is required for type with serial name

Привет, проблема о которой ты пишешь происходит потому, что если код не верный, то тебе приходит вот такой ответ

```json
{
  "remaining_attempts": 0,
  "lock_duration": 0,
  "error": "Invalid token for the last code",
  "status_code": 401
}

```
а если верный то вот такой

```json
{
  "message": "Successfully confirmed email",
  "status_code": 200
}

```

ты ожидаешь, что в ответе будет поле message, но его нет.
я рекомендую тебе прежде чем десериализовать **весь json** узнать какой у тебя статус код
например ты можешь десериализовать ответ в класс у которого есть только одно поле — статус код.
На основе полученного статус кода уже решать в какой объект совершать десериализацию.

MajoritySky2496 commented:
@artwist-polyakov в том то и проблема что я это сделать) ошибка должна обработаться в dataSource как NetworkException, но она ловится раньше в репозитории как обычный exception. Когда дебажил, пару раз обрабатывалось правильно, но в основном какая то чертовщина

artwist-polyakov commented:
> @artwist-polyakov в том то и проблема что я это сделать) ошибка должна обработаться в dataSource как NetworkException, но она ловится раньше в репозитории как обычный exception. Когда дебажил, пару раз обрабатывалось правильно, но в основном какая то чертовщина

привет.

Ну как же чертовщина.
Давай еще раз посмотрим на код.
А конкретно вот на этот фрагмент

![CleanShot 2024-05-11 at 11 50 12@2x](https://github.com/CashAdv/CashADV/assets/42345388/913024e2-5732-47e1-8c14-57b5729b4451)

Нас интересует вот этот класс `ResetPasswordConfirmationResponse` (сорри, тут я не совсем то копирнул, но схема в целом понятна)

```kotlin

@Serializable
data class ResetPasswordConfirmationResponse(
    val message:String,
    val token:String,
    @SerialName("status_code") val statusCode: Int = 0)

```
именно его ты получаешь в ответ верификацию кода.
там у тебя есть `message` — но его на самом деле нет в ответе.

Я беру, заменяю этот кусок кода вот на такое (параллельно распаковывая опционал в конверторе):

```kotlin

@Serializable
data class ConfirmResetPasswordResponse(
    val message: String? = null,
    @SerialName("status_code") val statusCode: Int = 0
)

```

и тадам — меня пропускает на следующий шаг ввода пароля не смотря на то что статус код неверный. и сам код тоже не верный.
**отсюда я делаю вывод что ты не проверяешь код ответа, который хранится в результате ответа на твой запрос.**

Вообще обрати внимание, что когда ты начинаешь работать с исключениями, то есть смысл также отрабатывать "остальной" кейс — когда ты попадаешь не в предусмотренные программой NetworkExceptions, а в какие-то другие, как в твоем случае.
Тогда можно было бы выводить понятный лог и как то пояснять, что всё же случилось, но у тебя там была пустота, которая никак не логировалась. 

PS. И еще кое-что.
Попробуй называть классы как-то очень понятно, чтобы сходу было ясно за что они отвечают.
Сейчас вот например `ConfirmResetPasswordResponse` и `ResetPasswordConfirmationResponse` кажется не отличаются ничем)))

PS. Обрати внимание на длины строк, вроде как 100 симв оптимально для работы, постарайся не делать их такими огромными; я пока смотрел находил примеры более 150 симв длиной

MajoritySky2496 committed:
71-Password-Recovery
1. Solved the "empty message exception"
2. Renamed the class names to more understandable ones

MajoritySky2496 committed:
Merge branch 'dev' into 71-Password-Recovery

MajoritySky2496 committed:
71-Password-Recovery
1. Resolved conflicts

MajoritySky2496 committed:
71-Password-Recovery
1. Made the lines shorter

artwist-polyakov reviewed (changes requested):
Пара замечаний:

1. Если мы вводим неправильный пароль, то вместе с сообщением о том, что пароль не правильный необходимо также очищать поле ввода. У вьюхи кода есть для этого метод.
2. Может быть сделаем тоаст в котором приведем текущий код подтверждения прям в приложении? чтобы не ходить за ним далеко ?

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/presentation/viewmodel/PasswordRecoveryViewModel.kt:
```diff
@@ -0,0 +1,437 @@
+
+}
+
+fun setEmailConfirmCode(code: String, context: Context) {
```
передавать контекст в ViewModel запрещенно

MajoritySky2496 replied:
конекст убрал

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/presentation/viewmodel/PasswordRecoveryViewModel.kt:
```diff
@@ -0,0 +1,437 @@
+
+
+
+     fun setEmail(
```
название setEmail наводит на мысль что будет установка e-mail, а по факту выполняется запрос на восстановление пароля
тогда стоит назвать метод requestRecovery или tryToRecovery

при этом у тебя есть состояние инпута емейла в emailInput, можно сюда его дополнительно не передавать

MajoritySky2496 replied:
исправил

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/presentation/viewmodel/PasswordRecoveryViewModel.kt:
```diff
@@ -0,0 +1,437 @@
+
+private fun recovery() {
+
+    viewModelScope.launch {
+        resetPasswordInteractor.isResetPasswordInProgress().collect{
+            _uiState.value
+        }
+    }
```
а что эта подписка делает?

MajoritySky2496 replied:
Подсмотрел у ребят, но так и не нашел ей применение, убрал ее

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/presentation/viewmodel/PasswordRecoveryViewModel.kt:
```diff
@@ -0,0 +1,437 @@
+            is Resource.Success -> {
+                viewModelScope.launch {
+                    logDebugMessage(result.data)
+                    resendCountDownJob!!.cancel()
```
нельзя вызывать метод через !! - будет краш

MajoritySky2496 replied:
исправил

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/presentation/viewmodel/PasswordRecoveryViewModel.kt:
```diff
@@ -0,0 +1,437 @@
+            }
+
+            is Resource.Error -> {
+                when (result.error) {
+                    is ErrorEntity.SaveNewPassword.InvalidInput -> {
+                        logDebugMessage("InvalidInput ${result.error.message}")
+                        _messageEvent.emit(
+                            RecoveryScreenMessageContent.ResetPasswordError(result.error.message))
+                    }
+
+                    is ErrorEntity.SaveNewPassword.InvalidToken -> {
+                        logDebugMessage("InvalidToken ${result.error.message}")
+                        _messageEvent.emit(RecoveryScreenMessageContent
+                            .ResetPasswordError(result.error.message))
+                    }
+
+                    is ErrorEntity.SaveNewPassword.FailedToResetPassword -> {
+                        logDebugMessage("FailedToResetPassword ${result.error.message}")
+                        _messageEvent.emit(
+                            RecoveryScreenMessageContent.ResetPasswordError(result.error.message))
+                    }
+
+                    is ErrorEntity.NetworksError.NoInternet -> {
+                        logDebugMessage("NoInternet ${result.error.message}")
+                        _sideEffects.emit(RecoverySideEffect.NoInternetConnection)
+                    }
+
+                    else -> {
+                        logDebugMessage("Something went wrong ${result.error.message}")
+                        _messageEvent.emit(
+                            RecoveryScreenMessageContent.ResetPasswordError(result.error.message))
+                    }
+
+                }
```
logDebugMessage можно написать один раз

MajoritySky2496 replied:
исправил

avanisimov commented:
Отформатируй, пожалуйста, свой код. сейчас его сложно читать из-за кривой табуляции

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/data/impl/ResetPasswordRepositoryImpl.kt:
```diff
@@ -0,0 +1,198 @@
+    private val _state: MutableStateFlow<ResetPasswordState> =
+        MutableStateFlow(ResetPasswordState())
+    private val state = _state.asStateFlow()
+    private val currentState: ResetPasswordState
+        get() = state.replayCache.firstOrNull()
+            ?: ResetPasswordState(ResetPasswordState.State.Initial)
```
private val currentState: ResetPasswordState
        get() = _state.value

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/domain/ResetDomainMapper.kt:
```diff
@@ -0,0 +1,60 @@
+
+class ResetDomainMapper @Inject constructor() {
+
+    fun toResetPasswordInputDto(email: Email): ResetPasswordInputDto {
```
домейн слой не должен знать про существование DTO

MajoritySky2496 replied:
исправил

MajoritySky2496 committed:
71-Password-Recovery
1) removed context from viewmodel
2) added Resources in viewmodel
3) renamed fun setEmail(email:String) on requestRecovery()
4) removed isResetPasswordInProgress()
5)I removed the extra one logDebugMessage
6)formatted the code
7) removed DTO from Domain
8) the code field is cleared when an error occurs

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/presentation/ui/PasswordRecoveryFragment.kt:
```diff
@@ -1,45 +1,384 @@
+    private fun navigateBack(state: PasswordRecoveryScreenState) {
+        with(binding) {
+            when (state) {
+                is PasswordRecoveryScreenState.ConfirmationCode -> {
+                    btnBack.setOnClickListener {
+                        viewModel.navigateBackToEmailState()
+                    }
+
+                    requireActivity().onBackPressedDispatcher.addCallback {
+                        viewModel.navigateBackToEmailState()
+                    }
+
+                }
 
-    private var _binding: FragmentPasswordRecoveryBinding? = null
-    private val binding get() = _binding!!
+                is PasswordRecoveryScreenState.EmailInput -> {
+                    binding.btnBack.setOnClickListener {
+                        findNavController().navigateUp()
+                    }
+
+                    requireActivity().onBackPressedDispatcher.addCallback {
+                        findNavController().navigateUp()
+                    }
+                }
+
+                is PasswordRecoveryScreenState.PasswordInput -> {
+                    binding.btnBack.setOnClickListener {
+                        viewModel.navigateBackToEmailState()
+                    }
+                    requireActivity().onBackPressedDispatcher.addCallback {
+                        viewModel.navigateBackToEmailState()
```
1 - сейчас у тебя логика "как обрабатывать кнопку назад?" находится по факту во фрагменте
2 - requireActivity().onBackPressedDispatcher.addCallback вызывается много раз и в итоге много колбеков накапливается в диспатчере

лучше один раз в начале жизни фрагмента прокинуть onBackPressed внутрь вьюмодели, а вот он уже сделает логику when (state) {....} и сайд-эффектами будет управлять фрагментом

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/presentation/ui/PasswordRecoveryFragment.kt:
```diff
@@ -1,45 +1,384 @@
+    private fun showSnackbar(message: String, viewToFocus: TextInputEditText) {
+        hideKeyboard()
+        Snackbar.make(binding.root, message, LoginFragment.SNACKBAR_DURATION)
+            .setBackgroundTint(resources.getColor(R.color.black, null))
+            .setTextColor(resources.getColor(R.color.white, null))
+            .setActionTextColor(resources.getColor(R.color.white, null))
```
а чего тему не передаёшь?

MajoritySky2496 replied:
не понимал зачем это нужно, теперь разобрался

artwist-polyakov reviewed (approved):
У меня замечаний нет (ну кроме идеи про код в тоасте, которая не может быть основанием не апрувить ревью)

MajoritySky2496 committed:
71-Password-Recovery
1) Called onBackPressedDispatcher.addCallback in viewmodel
2) added a theme
